File: engines/typescript/src/assess.test.ts
/**
 * The full assess path (graph → catalog → structural → assertions) is byte-identical to the reference.
 *
 * `testdata/assess-golden.json` is the reference engine's `assess_subjects` output over the base
 * catalog for several event sets (a control's passed and failed fixtures, and an empty stream). The
 * comparison also covers evidence pointers, whose digests canonicalise the real events — so any drift
 * in the canonical form, the graph, the shape parse, or the evaluator would break it.
 */

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";
import { assertionToJson } from "./assertions";
import { assessSubjects } from "./assess";
import { canonicalString } from "./canonical";
import { loadCatalog } from "./catalog";
import { DomainBinding } from "./domain";
import { profileFromDict } from "./profile";

const REPO = join(__dirname, "..", "..", "..");
const BASE = join(REPO, "spec", "catalogs", "base", "eu-ai-act");
const TESTDATA = join(__dirname, "..", "testdata");
const SUBJECT = "spiffe://corp/agents/a";

function eventsOf(controlId: string, kase: string): Array<Record<string, unknown>> {
  return readFileSync(join(BASE, "test", controlId, `${kase}.jsonl`), "utf-8")
    .split("\n")
    .filter((line) => line.trim())
    .map((line) => JSON.parse(line));
}

test("assess_subjects over the base catalog matches the Python reference", () => {
  const golden = JSON.parse(readFileSync(join(TESTDATA, "assess-golden.json"), "utf-8"));
  const catalog = loadCatalog(BASE);
  const domain = DomainBinding.load(join(BASE, "test", "domain.yaml"));
  const profile = profileFromDict({
    subjects: [{ id: SUBJECT, role: "both" }],
    catalogs: ["base/eu-ai-act"],
  });

  const scenarios: Record<string, Array<Record<string, unknown>>> = {
    "ovs-passed": eventsOf("OVS-03", "passed"),
    "ovs-failed": eventsOf("OVS-03", "failed"),
    "rec-passed": eventsOf("REC-04", "passed"),
    empty: [],
  };

  const result: Record<string, unknown> = {};
  for (const [name, events] of Object.entries(scenarios)) {
    result[name] = assessSubjects(events, profile, [catalog], domain).map(assertionToJson);
  }
  assert.equal(canonicalString(result), canonicalString(golden));
});

File: engines/java/src/test/java/org/agentce/AssessTest.java
package org.agentce;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * The full assess path (graph → catalog → structural → assertions) is byte-identical to the reference.
 * {@code assess-golden.json} is the reference engine's {@code assess_subjects} output over the base
 * catalog for several event sets; the comparison covers evidence-pointer digests, so any drift in the
 * canonical form, graph, shape parse, or evaluator would break it.
 */
class AssessTest {
    private static final String SUBJECT = "spiffe://corp/agents/a";

    private static List<JsonNode> eventsOf(String controlId, String kase) throws IOException {
        return Fixtures.readJsonl(Fixtures.BASE.resolve("test").resolve(controlId).resolve(kase + ".jsonl"));
    }

    @Test
    void assessSubjectsMatchesReference() throws IOException {
        JsonNode golden = Json.parseFile(TestPaths.testData().resolve("assess-golden.json"));
        Catalog catalog = Catalog.load(Fixtures.BASE);
        DomainBinding domain = DomainBinding.load(Fixtures.BASE.resolve("test/domain.yaml"));
        Profile profile = Profile.fromDict(
                Json.parse("{\"subjects\":[{\"id\":\"" + SUBJECT + "\",\"role\":\"both\"}],\"catalogs\":[\"base/eu-ai-act\"]}"));

        Map<String, List<JsonNode>> scenarios = new LinkedHashMap<>();
        scenarios.put("ovs-passed", eventsOf("OVS-03", "passed"));
        scenarios.put("ovs-failed", eventsOf("OVS-03", "failed"));
        scenarios.put("rec-passed", eventsOf("REC-04", "passed"));
        scenarios.put("empty", List.of());

        ObjectNode result = Json.nodes().objectNode();
        for (Map.Entry<String, List<JsonNode>> e : scenarios.entrySet()) {
            ArrayNode arr = result.putArray(e.getKey());
            for (Assertions.Assertion a : Assess.assessSubjects(e.getValue(), profile, List.of(catalog), domain)) {
                arr.add(a.toJson());
            }
        }
        assertEquals(Canonical.canonicalString(golden), Canonical.canonicalString(result));
    }
}
